File: 2020/aoc-clojure-2020/src/aoc_clojure_2020/day-07.clj
(ns aoc-clojure-2020.day-07
  (:require [aoc-clojure-2020.helpers :refer [get-input]]
            [clojure.string :as s]))

(defn parse-line [l]
  (let [[_ colour contained] (re-matches #"([a-z\ ]+) bags contain (.+)" l)
        contained-cols (into {} (map (fn [[_ n c]] [c (Integer/parseInt n)]))
                             (re-seq #"(\d+) ([a-z\ ]+) bag" contained))]
    [colour contained-cols]))

(def input (->> (get-input 7)
                s/split-lines
                (into {} (map parse-line))))

(defn bag-contains [bag contained rules]
  (if (contains? (get rules bag) contained)
    true
    (some #(bag-contains (first %) contained rules) (get rules bag))))

(defn count-bag [bag rules]
  (reduce (fn [acc [col n]] (+ acc (* n (inc (count-bag col rules))))) 0 (get rules bag)))

(def part-1 (-> (filter #(bag-contains (first %) "shiny gold" input) input) count))
(def part-2 (count-bag "shiny gold" input))

;; iterative solution =================

(defn count-bags-containing-child [bag rules]
  (->> (loop [to-check (keys rules)
              checked {}]
         (if-let [c (first to-check)]
           (if (contains? checked c)
             (recur (rest to-check) checked)
             (if-let [children (keys (get rules c))]
               (if (some #(= % bag) children)
                 (recur (rest to-check) (conj checked [c true]))
                 (if-let [to-calculate (not-empty (filter #(nil? (checked %)) children))]
                   (recur (apply conj to-check to-calculate) checked)
                   (recur (rest to-check) (conj checked [c (boolean (some #(checked %) children))]))))
               (recur (rest to-check) (conj checked [c false]))))
           checked))
       (filter second)
       count))

(def part-1-iter (count-bags-containing-child "shiny gold" input))

File: 2020/aoc-elixir/lib/advent_of_code/day_07.ex
defmodule AdventOfCode.Day07 do
  
  def parse_line(line) do
    [colour, contained] = Regex.run(~r/([[:alnum:]\ ]+) bags contain (.+)/,
      line, capture: :all_but_first)
    {colour, Regex.scan(~r/(\d+) ([[:alnum:]\ ]+) bag/, contained, capture: :all_but_first)
    |> Enum.map(fn [n, c] -> {c, String.to_integer(n)} end)
    |> Map.new}
  end

  def parse_input(input) do
    input |> String.trim |> String.split("\n") |> Enum.map(&parse_line/1) |> Map.new
  end

  def contains_shiny_gold(colour, rules) do
    if Map.has_key?(rules[colour], "shiny gold") do
      true
    else
      Enum.any?(rules[colour], fn {c, _} -> contains_shiny_gold(c, rules) end)
    end
  end

  def count_bags(bag, rules) do
    Enum.reduce(rules[bag], 0, fn {c, n}, acc -> acc + (n * (count_bags(c, rules) + 1)) end)
  end
  
  def part1(args) do
    input = parse_input(args)
    Enum.filter(input, fn {c, _} -> contains_shiny_gold(c, input) end)
    |> Enum.count
  end

  def part2(args), do: count_bags("shiny gold", parse_input(args))
end
